pkg/controller/volume/persistentvolume/pv_controller_base.go

```diff
@@ -428,29 +468,58 @@ func updateMigrationAnnotationsAndFinalizers(cmpm CSIMigratedPluginManager, tran
 		}
 		return modified
 	} else {
+		// Migration Disabled
+		annModified := false
+		finalizerModified := false
 		if migratedToDriver != "" {
 			// Migration annotation exists but the driver isn't migrated currently
 			delete(ann, storagehelpers.AnnMigratedTo)
-			if !claim && utilfeature.DefaultFeatureGate.Enabled(features.HonorPVReclaimPolicy) {
-				modified := false
-				if finalizers == nil {
-					finalizers = &([]string{})
-				}
-				// Add back the in-tree PV deletion protection finalizer if does not already exists
-				if !slice.ContainsString(*finalizers, storagehelpers.PVDeletionInTreeProtectionFinalizer, nil) {
-					*finalizers = append(*finalizers, storagehelpers.PVDeletionInTreeProtectionFinalizer)
-					modified = true
-				}
-				// Remove the external PV deletion protection finalizer
-				if slice.ContainsString(*finalizers, storagehelpers.PVDeletionProtectionFinalizer, nil) {
-					*finalizers = slice.RemoveString(*finalizers, storagehelpers.PVDeletionProtectionFinalizer, nil)
-					modified = true
-				}
-				return modified
-			}
+			annModified = true
+		}
+		if !claim {
+			// Applicable to all dynamically provisioned volumes.
+			// This ensures that the finalizer is added on all pre-existing volumes.
+			finalizerModified = modifyVolumeFinalizers(finalizers, pvFinalizersModSpec.pvReclaimPolicy, pvFinalizersModSpec.isInTreeVolume)
+		}
+		return annModified || finalizerModified
+	}
+}
+
+// modifyVolumeFinalizers updates the finalizers based on the reclaim policy and if it is a in-tree volume or not.
+// The in-tree PV deletion protection finalizer is only added if the reclaimPolicy associated with the PV is `Delete`.
+// The in-tree PV deletion protection finalizer is removed if the reclaimPolicy associated with the PV is `Retain` or
+// `Recycle`, removing the finalizer is necessary to reflect the recalimPolicy updates on the PV.
+// The method also removes any external PV Deletion Protection finalizers added on the PV, this represents CSI migration
+// rollback/disable scenarios.
+func modifyVolumeFinalizers(finalizers *[]string, reclaimPolicy v1.PersistentVolumeReclaimPolicy, isInTreeVolume bool) bool {
+	modified := false
+	if utilfeature.DefaultFeatureGate.Enabled(features.HonorPVReclaimPolicy) {
+		if finalizers == nil {
+			finalizers = &([]string{})
+		}
```

Comment:
This will modify the finalizers even when the function returns `false`.

What about:
```
var inFinalizers []string

if finalizers != nil {
    inFinalizers = *finalizers
}
```

Then use `inFinalizers` for all decissions and `*finalizers` only for the output (`*finalizers = append(inFinalizers, ...)` etc)